File: mcp_server.py
import asyncio
import json
from query_test_cdc import fetch_test_cdc_rows
from export_to_csv import export_test_cdc_to_csv
import io
import sys
from contextlib import redirect_stdout
import os

def debug_print(msg):
    print(msg, file=sys.stderr, flush=True)

class MCPServer:
    def __init__(self, host='localhost', port=1337):
        self.host = host
        self.port = port
        self.handlers = {
            'query_test_cdc': self.handle_query,
            'export_to_csv': self.handle_export
        }
        debug_print(f"Initializing MCP Server on {host}:{port}")

    async def send_message(self, writer, message):
        try:
            message_bytes = json.dumps(message).encode()
            writer.write(len(message_bytes).to_bytes(4, 'big'))
            writer.write(message_bytes)
            await writer.drain()
            debug_print(f"Sent message: {message}")
        except Exception as e:
            debug_print(f"Error sending message: {e}")
            raise

    async def handle_client(self, reader, writer):
        try:
            while True:
                # Read the length prefix (4 bytes)
                length_bytes = await reader.read(4)
                if not length_bytes:
                    debug_print("Client disconnected (no length bytes)")
                    break

                message_length = int.from_bytes(length_bytes, 'big')
                debug_print(f"Received message length: {message_length}")

                # Read the message
                message_bytes = await reader.read(message_length)
                if not message_bytes:
                    debug_print("Client disconnected (no message bytes)")
                    break

                # Parse the message
                message = json.loads(message_bytes.decode())
                debug_print(f"Received message: {message}")

                # Handle initialization specially
                if message.get('method') == 'initialize':
                    # Send immediate response
                    init_response = {
                        'jsonrpc': '2.0',
                        'id': message.get('id'),
                        'result': {
                            'protocolVersion': message['params']['protocolVersion'],
                            'capabilities': {}
                        }
                    }
                    await self.send_message(writer, init_response)

                    # Send ready notification
                    ready_notification = {
                        'jsonrpc': '2.0',
                        'method': 'server/ready',
                        'params': {}
                    }
                    await self.send_message(writer, ready_notification)
                    continue

                # Handle commands
                if message.get('command'):
                    command = message.get('command')
                    if command not in self.handlers:
                        error_response = {
                            'jsonrpc': '2.0',
                            'id': message.get('id'),
                            'error': {
                                'code': -32601,
                                'message': f'Unknown command: {command}'
                            }
                        }
                        await self.send_message(writer, error_response)
                        continue

                    handler = self.handlers[command]
                    try:
                        result = await handler(message)
                        await self.send_message(writer, result)
                    except Exception as e:
                        error_response = {
                            'jsonrpc': '2.0',
                            'id': message.get('id'),
                            'error': {
                                'code': -32000,
                                'message': str(e)
                            }
                        }
                        await self.send_message(writer, error_response)

        except Exception as e:
            debug_print(f"Error in handle_client: {e}")
            try:
                error_response = {
                    'jsonrpc': '2.0',
                    'id': message.get('id') if 'message' in locals() else None,
                    'error': {
                        'code': -32000,
                        'message': str(e)
                    }
                }
                await self.send_message(writer, error_response)
            except:
                pass
        finally:
            writer.close()
            try:
                await writer.wait_closed()
            except:
                pass
            debug_print("Client connection closed")

    async def handle_query(self, message):
        debug_print("Executing query handler")
        output = io.StringIO()
        with redirect_stdout(output):
            try:
                fetch_test_cdc_rows()
                return {
                    'jsonrpc': '2.0',
                    'id': message.get('id'),
                    'result': {
                        'output': output.getvalue()
                    }
                }
            except Exception as e:
                debug_print(f"Error in query handler: {e}")
                return {
                    'jsonrpc': '2.0',
                    'id': message.get('id'),
                    'error': {
                        'code': -32000,
                        'message': str(e)
                    }
                }

    async def handle_export(self, message):
        debug_print("Executing export handler")
        output = io.StringIO()
        with redirect_stdout(output):
            try:
                export_test_cdc_to_csv()
                return {
                    'jsonrpc': '2.0',
                    'id': message.get('id'),
                    'result': {
                        'output': output.getvalue()
                    }
                }
            except Exception as e:
                debug_print(f"Error in export handler: {e}")
                return {
                    'jsonrpc': '2.0',
                    'id': message.get('id'),
                    'error': {
                        'code': -32000,
                        'message': str(e)
                    }
                }

    async def start(self):
        try:
            server = await asyncio.start_server(
                self.handle_client,
                self.host,
                self.port
            )
            
            debug_print(f"MCP Server running on {self.host}:{self.port}")
            
            async with server:
                await server.serve_forever()
        except Exception as e:
            debug_print(f"Error starting server: {e}")
            raise

if __name__ == "__main__":
    debug_print("Starting MCP Server...")
    server = MCPServer()
    try:
        asyncio.run(server.start())
    except KeyboardInterrupt:
        debug_print("\nServer shutdown requested")
    except Exception as e:
        debug_print(f"Server error: {e}")
    finally:
        debug_print("Server shutdown complete") 

File: mcp_server.js
#!/usr/bin/env node

const oracledb = require('oracledb');
require('dotenv').config();

// Initialize Oracle client
oracledb.initOracleClient({ libDir: process.env.ORACLE_HOME });

class MCPServer {
    constructor() {
        this.tools = [
            {
                name: "query_test_cdc",
                description: "Query the first 10 rows from FDW database TEST_CDC table",
                inputSchema: {
                    type: "object",
                    properties: {},
                    required: []
                }
            },
            {
                name: "export_to_csv",
                description: "Export TEST_CDC data from FDW database to CSV file",
                inputSchema: {
                    type: "object",
                    properties: {},
                    required: []
                }
            }
        ];
    }

    debug(msg) {
        console.error(`[DEBUG] ${msg}`);
    }

    async getOracleConnection() {
        return await oracledb.getConnection({
            user: process.env.ORACLE_USER,
            password: process.env.ORACLE_PASSWORD,
            connectString: process.env.ORACLE_DSN
        });
    }

    async handleQuery() {
        try {
            const connection = await this.getOracleConnection();
            const result = await connection.execute(
                'SELECT * FROM TEST_CDC WHERE ROWNUM <= 10',
                [],
                { outFormat: oracledb.OUT_FORMAT_OBJECT }
            );
            await connection.close();

            return {
                content: [
                    {
                        type: "text",
                        text: `Query Results from FDW database TEST_CDC table:\n\n${JSON.stringify(result.rows, null, 2)}`
                    }
                ]
            };
        } catch (error) {
            this.debug(`Query error: ${error.message}`);
            return {
                content: [
                    {
                        type: "text",
                        text: `Error querying FDW database: ${error.message}`
                    }
                ],
                isError: true
            };
        }
    }

    async handleExport() {
        try {
            const connection = await this.getOracleConnection();
            const result = await connection.execute(
                'SELECT * FROM TEST_CDC WHERE ROWNUM <= 10',
                [],
                { outFormat: oracledb.OUT_FORMAT_OBJECT }
            );
            await connection.close();

            if (result.rows.length === 0) {
                return {
                    content: [
                        {
                            type: "text",
                            text: "No data found in FDW database TEST_CDC table to export."
                        }
                    ]
                };
            }

            const fs = require('fs');
            const filename = 'test_cdc_export.csv';
            
            // Convert to CSV
            const headers = Object.keys(result.rows[0]).join(',');
            const rows = result.rows.map(row => 
                Object.values(row).map(val => 
                    typeof val === 'string' ? `"${val.replace(/"/g, '""')}"` : val
                ).join(',')
            );
            
            fs.writeFileSync(filename, [headers, ...rows].join('\n'));

            return {
                content: [
                    {
                        type: "text",
                        text: `Data exported successfully from FDW database to ${filename}\n\nExported ${result.rows.length} rows with columns: ${headers}`
                    }
                ]
            };
        } catch (error) {
            this.debug(`Export error: ${error.message}`);
            return {
                content: [
                    {
                        type: "text",
                        text: `Error exporting data from FDW database: ${error.message}`
                    }
                ],
                isError: true
            };
        }
    }

    async handleRequest(request) {
        const { method, params } = request;

        switch (method) {
            case 'initialize':
                return {
                    protocolVersion: params.protocolVersion,
                    capabilities: {
                        tools: {},
                        resources: {}
                    },
                    serverInfo: {
                        name: "fdw-mcp-server",
                        version: "1.0.0"
                    }
                };

            case 'notifications/initialized':
                // This is a notification, no response needed
                return null;

            case 'tools/list':
                return {
                    tools: this.tools
                };

            case 'resources/list':
                return {
                    resources: []
                };

            case 'tools/call':
                const { name, arguments: args } = params;
                
                switch (name) {
                    case 'query_test_cdc':
                        return await this.handleQuery();
                    case 'export_to_csv':
                        return await this.handleExport();
                    default:
                        throw new Error(`Unknown tool: ${name}`);
                }

            default:
                throw new Error(`Unknown method: ${method}`);
        }
    }

    sendResponse(id, result) {
        if (result !== null) {
            const response = {
                jsonrpc: "2.0",
                id: id,
                result: result
            };
            console.log(JSON.stringify(response));
        }
    }

    sendError(id, error) {
        const response = {
            jsonrpc: "2.0",
            id: id,
            error: {
                code: -32000,
                message: error.message
            }
        };
        console.log(JSON.stringify(response));
    }

    start() {
        this.debug('FDW MCP Server starting...');

        process.stdin.setEncoding('utf8');
        
        let buffer = '';
        
        process.stdin.on('data', async (chunk) => {
            buffer += chunk;
            
            // Process complete lines
            const lines = buffer.split('\n');
            buffer = lines.pop(); // Keep incomplete line in buffer
            
            for (const line of lines) {
                if (line.trim()) {
                    let currentRequest = null;
                    try {
                        currentRequest = JSON.parse(line);
                        this.debug(`Received request: ${JSON.stringify(currentRequest)}`);
                        
                        const result = await this.handleRequest(currentRequest);
                        this.sendResponse(currentRequest.id, result);
                        
                    } catch (error) {
                        this.debug(`Error processing request: ${error.message}`);
                        this.sendError(currentRequest?.id || null, error);
                    }
                }
            }
        });

        process.stdin.on('end', () => {
            this.debug('FDW MCP Server shutting down...');
            process.exit(0);
        });

        this.debug('FDW MCP Server ready for requests');
    }
}

const server = new MCPServer();
server.start(); 